### ISSUE TITLE
MaskedArray heuristic for memory overlap seems simplistic and slow

### ISSUE THREAD

**Issue by mhvk** [numpy/numpy#10234]
Currently, in `MaskedArray.__array_finalize__`, the [following](https://github.com/numpy/numpy/blob/master/numpy/ma/core.py#L2981) is done to check whether the new object may be a view of an old one:
```
if (obj.__array_interface__["data"][0] != self.__array_interface__["data"][0]):
```
if that check fails, the mask is copied.

This seems unnecessarily restrictive and fails even simple slicing:
```
ma = np.ma.MaskedArray(np.arange(100.))
ma2 = ma[10:20]
ma.__array_interface__["data"][0] == ma2.__array_interface__["data"][0]
# False
```
This means that slices by default do not share the mask with the original object, which doesn't seem a good idea given that we tried to change this behaviour in #5580 (hence, cc @jakirkham).

A relatively straightforward solution would be to replace it with `not np.may_share_memory(ma, ma2)`
(this, perhaps surprisingly, is also much faster than the above for the simple slice case).

**Comment by eric-wieser** [comment]
Are you sure that slicing isn't already handled correctly by `__getitem__`?

**Comment by mhvk** [comment]
Yes, you're right, it is. I guess that if we changed the behaviour here, some of the work-arounds in `__getitem__` could be removed. But maybe it is not worth the hassle (though the present comparison takes 4us on my computer, while `may_share_memory` takes 0.3us)

**Comment by eric-wieser** [comment]
Most of the __getitem__ workaround are due to;
* Decay into scalars when indexing
* Not being able to reliably inspect the indexing tuple due to #9686
